migrations/educandu-2022-07-27-01-use-percentages-in-interactive-media-plugin.js

```diff
@@ -0,0 +1,76 @@
+/* eslint-disable camelcase, no-console, no-await-in-loop */
+
+import by from 'thenby';
+
+export default class Educandu_2022_07_27_01_use_percentages_in_interactive_media_plugin {
+  constructor(db) {
+    this.db = db;
+  }
+
+  round(num) {
+    return parseFloat(num.toFixed(15));
+  }
+
+  timecodeToPercentage(timecode, duration, fallbackValue) {
+    return Math.max(0, Math.min(1, duration ? this.round(timecode / duration) : fallbackValue));
+  }
+
+  processSectionContent(content) {
+    const { sourceDuration, sourceStartTimecode, sourceStopTimecode, chapters } = content;
+
+    content.playbackRange = [
+      this.timecodeToPercentage(sourceStartTimecode ?? 0, sourceDuration, 0),
+      this.timecodeToPercentage(sourceStopTimecode ?? sourceDuration, sourceDuration, 1)
+    ];
+
+    const playbackDuration = (content.playbackRange[1] - content.playbackRange[0]) * sourceDuration;
+
+    content.chapters = chapters
+      .map((chapter, index) => {
+        chapter.startPosition = this.timecodeToPercentage(chapter.startTimecode, playbackDuration, index / chapters.length);
+        delete chapter.startTimecode;
+        return chapter;
+      })
+      .sort(by(chapter => chapter.startPosition));
+
+    delete content.sourceDuration;
+    delete content.sourceStartTimecode;
+    delete content.sourceStopTimecode;
+    delete content.startTimecode;
+    delete content.stopTimecode;
+  }
+
+  async processCollection(collectionName) {
+    const toUpdate = await this.db.collection(collectionName).find({ 'sections.type': 'interactive-media' }).toArray();
+    let updateCount = 0;
+
+    for (const doc of toUpdate) {
+      let docWasUpdated = false;
+      for (const section of doc.sections) {
+        if (section.type === 'interactive-media' && section.content) {
+          this.processSectionContent(section.content);
+          docWasUpdated = true;
+          console.log(`Updating ${collectionName} ${doc._id} - section ${section.key}`);
+        }
+      }
+      if (docWasUpdated) {
+        updateCount += 1;
+        await this.db.collection(collectionName).replaceOne({ _id: doc._id }, doc);
+      }
+    }
+
+    return updateCount;
+  }
+
+  async up() {
+    const lessonsCount = await this.processCollection('lessons');
```

Comment:
I think it's safe to say you could drop this lessons migration now.

Reply:
True, dropped!